### PR TITLE
perf(encoder): wire donor-shape Fast kernel into MatchGeneratorDriver (#198 phase 1b)

### LINKED ISSUE #198: perf(fast): port donor ZSTD_compressBlock_fast — 22× regression on Fast strategy
## Problem

Compression speed on Fast strategy (levels -7 to 1) is catastrophically below donor:
- Rust: ~30 MB/s flat across all levels
- FFI: ~660 MB/s on L=-7, ~280 on L=-1, ~120 on L=1
- **Gap: 22× on L=-7**

Decode side on Fast levels also shows a 4.6× gap (~400 MB/s Rust vs ~1850 MB/s FFI) — likely related: Fast-level frames produce simple sequence streams whose decode is dominated by the RingBuffer wrap dispatch (see #178 backlog item #132 flat-buffer mode).

## Root cause (encode side)

Our `MatchGenerator::next_sequence` (`zstd/src/encoding/simple/mod.rs`) uses a fundamentally different algorithm than the donor:

| | Ours | Donor (zstd_fast.c) |
|---|---|---|
| Hash table | `SuffixStore` (hashmap-like, byte-slice keys) | `U32* hashTable` (direct index, single u32 slot) |
| Lookup | `for match_entry in self.window.iter() { ... suffixes.get(key) }` | `hashTable[hash0]` (single u32 load) |
| Match check | `common_prefix_len(...)` (Rust slice) | `MEM_read32(ip0) == MEM_read32(base + matchIdx)` |
| Pipelining | none (single ip cursor) | `ip0/ip1/ip2/ip3` lookahead pairs |
| Backward extension | n/a | `while ip0[-1] == match0[-1]` (1 ALU op + branch) |
| Forward extension | byte-by-byte | `ZSTD_count` (SIMD u64 XOR + trailing-zero) |

Each of these is independent of micro-optimisation work — even with all current decoder/encoder hot-path tweaks, the Rust Fast path will not approach donor throughput until the matcher itself is replaced with a direct port.

## Scope

Full rewrite of `encoding/simple/mod.rs` (and the corresponding entry in `match_generator.rs`) to mirror `ZSTD_compressBlock_fast_noDict_generic` (zstd_fast.c:192-456). Specifically:

- Replace `SuffixStore` with `Vec<u32>` of length `1 << hashLog`, indexed by `hashPtr(ip, hlog, mls)` (mls = 4 for fast, parameterise for L1/L2/L3 dispatch).
- Implement the `_start:` loop with `ip0/ip1/ip2/ip3` cursors, `step` / `nextStep` skip acceleration, and the repcode-at-ip2 / hash-at-ip0 / hash-at-ip1 schedule.
- Implement `MEM_read32` equivalent (`u32::from_le_bytes(slice.try_into().unwrap())` or unsafe `read_unaligned`) for 4-byte match probe.
- Implement `ZSTD_count` equivalent (8-byte XOR + leading-zero divide-by-8) for forward extension; share with the existing decoder side that already has SIMD u64 chunked compare.
- Backward extension byte-by-byte loop bounded by `anchor` / `prefixStart`.
- ip0 advance rules: on no-match `ip0 = ip2`, on match `ip0 = ip0 + mLength + literals`.
- `useCmov` switch (`ZSTD_match4Found_cmov` vs `_branch`) is a perf knob, port the branch variant first, add the cmov variant once the body is correct.

## Estimated effort

10–20 focused hours. Not a single-PR change — needs a series of commits with measurements after each phase:
1. New `Vec<u32>` hash table + minimal `ip0`-only loop with raw match check — should already close most of the gap.
2. Add `ip0/ip1` two-position lookahead.
3. Add `ip2/ip3` four-position pipeline + step skipping.
4. Backward extension.
5. Cmov match-found variant + mls dispatch (4/5/6/7).

## Related

- Decode-side companion: `#178` backlog item #132 (flat-buffer mode for DecodeBuffer) — will close most of the 4.6× decode gap on Fast levels once landed.

Part of #178.

### PR DESCRIPTION
**Pull request by polaz**
## Summary

Phase 1b of the #198 Fast strategy donor port — wires the donor-shape
kernel modules from phase 1a (#215 → merged in fc634644) into the
production hot path. The legacy SuffixStore-based `MatchGenerator`
that lived in `simple/mod.rs` is fully removed; `MatcherStorage::Simple`
now holds a `FastKernelMatcher` that drives `compress_block_fast<MLS>`
once per block.

## What changes

- **`simple/mod.rs`** collapses from 496 → 14 lines (only module
  declarations + docstring remain). `SuffixStore`, `WindowEntry`,
  `MatchGenerator`, `repcode_candidate`, `add_data`, `next_sequence`,
  `add_suffixes_till`, `insert_suffix_if_absent`,
  `add_suffixes_interleaved_fast`, `offset_match_len`, `reserve` —
  all gone with the wiring.
- **`simple/fast_matcher.rs`** is the new active matcher
  (introduced incrementally on this branch — see commit log). The
  full inherent surface required by the `Matcher` trait wiring:
  - `accept_data(space)` — stash a committed block in `pending`.
  - `start_matching(handler)` — extend history, dispatch the
    kernel on the just-appended block, forward every emitted
    `Sequence::Triple` to the handler plus the terminal
    `Sequence::Literals` from `tail_literals_len`.
  - `skip_matching_with_hint(hint)` — `None` / `Some(true)` =
    history append only; `Some(false)` = dictionary-priming path
    that ALSO pre-populates the hash table for every position in
    the appended range, so cross-block matches against
    dict-primed bytes work in subsequent blocks.
  - `trim_to_window()` — drop history past `max_window_size`, bump
    `prefix_start_index`, clear the hash table. Donor's
    `ZSTD_window_correctOverflow` equivalent. Returns evicted
    bytes for the dictionary-budget retire loop.
  - `last_committed_space()` — returns pending pre-processing OR
    `history[last_block_start..]` post-processing (legacy
    MatchGenerator `window.last().data` parity for the frame
    compressor's raw-block emission path).
  - `reset(window_log, hash_log, mls)` — per-frame state reset,
    keeps the hash table allocation if `(hash_log, mls)`
    unchanged.
- **`MatchGeneratorDriver`** Simple-arm wiring:
  - `commit_space` → `m.accept_data(space)`.
  - `start_matching::<Fast>` → `m.start_matching(handler)`.
  - `skip_matching_with_hint` → `m.skip_matching_with_hint(hint)`.
  - `reset` → `m.reset(params.window_log, FAST_LEVEL_1_HASH_LOG,
    FAST_LEVEL_1_MLS)`.
  - `prime_with_dictionary` → field-level `offset_hist` /
    `max_window_size` mutation continues to work (both opened to
    `pub(crate)` for legacy parity).
  - `trim_after_budget_retire` → `m.trim_to_window()`.
  - `suffix_pool` field removed (no consumer after the legacy
    matcher's SuffixStore reuse path went away).

## Invariants preserved

- `prefix_start_index >= 1` at all times. Position 0 in `history` is
  permanently sub-prefix so the hash table's empty-slot sentinel
  value `0` cannot be confused with a real match position.
  Pinned in `with_params` + `reset`.
- `history.len()` is bounded by `2 × max_window_size` post-append.
  Eviction in `extend_history_with_pending` enforces.
- `rep[0..2]` (kernel-tracked) ↔ `offset_hist[0..2]` (wire encoder-
  tracked) stay in sync via per-Triple `encode_offset_with_history`
  calls. The `lit_len == 0` edge case (donor's `rep[0]-1` shift)
  is not modeled today — the phase-1 kernel doesn't emit
  `lit_len == 0` Triples. A future cmov / lookahead-pipelined
  kernel variant will need explicit handling.

## Tests

568 / 568 pass on the full workspace nextest suite:

- 16 unit tests on `FastKernelMatcher` itself (lifecycle, accept +
  start, skip flavors, dict-prime hash population, eviction,
  boundary cases at HASH_READ_SIZE = 8, rep ↔ offset_hist sync,
  prefix-eviction during dict-priming).
- 32 unit tests on the underlying kernel (donor-formula parity,
  prefix-filter, repcode backward extension three-piece proof,
  short-input early-return uniformity).
- All `frame_compressor` integration tests (raw-block detection,
  hinted source-size matrix, level-1 round-trips through both
  our decoder and FFI decode).
- All `cross_validation` Rust-encoded → FFI-decoded round-trips
  (every level 1..=22, dict + no-dict, encoded by Rust then
  read by the C reference decoder verbatim).

18 legacy tests in `match_generator.rs` that tested specifically
`SuffixStore` or `MatchGenerator::new(...)` behavior are
`#[cfg(any())]`-gated with explanatory comments — their substance
either has equivalent coverage in the new tests or relied on
algorithm-specific quirks of the SuffixStore matcher that the
donor-shape kernel doesn't reproduce by design (e.g. matching on
8-byte blocks below the kernel's `HASH_READ_SIZE = 8` floor).

## Benchmark (i9-9900K, `compress/level_1_fast/*/matrix/pure_rust`)

`cargo bench` deltas vs `main` (fc634644, phase 1a merged but not
wired) on the same host. Negative = faster, positive = slower.

| Scenario | Δ time | Throughput | Note |
|----------|-------:|-----------:|------|
| `low-entropy-1m` | **−90.7%** | 3.3 GiB/s | 10× faster (hot-path dominated by `count_forward`) |
| `decodecorpus-z000033` | **−83.4%** | 156 MiB/s | 6× faster (mixed corpus, donor-shape gets the typical case right) |
| `high-entropy-1m` | **−67.3%** | 633 MiB/s | 3× faster (no matches → kernel walks fast, exits early) |
| `small-10k-random` | **−36.0%** | 768 MiB/s | small random — kernel's step-skip helps |
| `small-4k-log-lines` | −15.4% | 154 MiB/s | small compressible — marginal |
| `small-1k-random` | +42.2% | 130 MiB/s | tiny block (1 KiB < HASH_READ_SIZE × 32 step) — kernel overhead exceeds work |
| `large-log-stream` | **+122.3%** | 235 MiB/s | **regression** — 25 MiB log; donor-parity `hash_log = 14` (16K slots) collides heavily on dense repeating content. See follow-up. |

Decompress side picks up a small indirect win (−3% to −5%) on
random-data scenarios because the sequence shape coming out of the
donor-shape kernel matches the decoder's optimised paths more
closely (fewer mid-block `Literals` emissions → less RingBuffer
dispatch).

### `large-log-stream` regression analysis

The regression is **expected** given the trade-off chosen at the
phase 1b kickoff (donor parity over speed):

- Legacy `SuffixStore` matcher used an effectively `window_log`-sized
  hash store (512K slots for level 1's `window_log = 19`) plus
  `hash_fill_step = 3` interleaving. Collision rate stayed low even
  on 25 MiB logs.
- Phase 1b uses the donor's level-1 `hash_log = 14` (16K slots).
  Against 25 MiB of compressible log content, the collision rate
  per slot climbs into the thousands, and the kernel's single-cursor
  loop loses matches the legacy matcher would have caught.
- Donor C zstd at level 1 Fast also pays this — the issue body's
  baseline shows donor at ~120 MB/s on similar workloads. The
  current main (this PR's baseline) was already faster than donor
  on this scenario because the legacy matcher was over-engineered
  for speed at the cost of `O(window)` memory.

Closing the gap is exactly the scope of issue #198 phase 3 (items
2 / 3 / 5 in the issue body): 4-cursor `ip0/ip1/ip2/ip3` lookahead
pipelining + `cmov` match-found variant + per-level `mls` dispatch.
Phase 3 lands on a follow-up branch once phase 1b is reviewed and
merged.

The net of phase 1b across the scenario matrix is positive
(geomean ≈ 2× faster, headline scenarios up to 10×). The
regression is bounded to dense-compressible large logs and is on
the documented roadmap.

## What's NOT in this PR (lands separately)

- Phase 1 → phase 3 follow-ups from #198: `ip0/ip1/ip2/ip3`
  4-cursor pipelining, `cmov` match-found variant, mls dispatch
  knob per level (`hash_fill_step` stride for dict-priming is
  still hard-coded to 1 today — the LevelParams field is wired
  in but the Fast matcher always strides at 1).
- `LevelParams.hash_log` / `LevelParams.mls` fields (currently
  Fast hard-codes donor level-1 defaults 14 / 7 from
  `FAST_LEVEL_1_HASH_LOG` / `FAST_LEVEL_1_MLS`). Per-level scaling
  for small-source-hint windows lands when `LevelParams` grows
  these fields.

Closes #198 phase 1b (the per-issue plan calls phase 3 a separate
follow-up branch).

Related: #178 (umbrella regression issue), #215 (phase 1a — kernel
modules, merged in fc634644).

## Summary by CodeRabbit

* **Refactor**
  * Replaced the legacy simple-match backend with a new fast‑kernel single‑pass matcher, streamlining lifecycle and memory reuse.
* **New Features**
  * Added a fast matcher with persistent recent‑match history, eager eviction on commit, and dictionary‑priming to improve match quality and throughput.
* **Bug Fixes**
  * Improved eviction, window trimming, and byte‑accounting during streaming commits for correct recycling and history handling.
* **Tests**
  * Disabled fragile legacy tests and relaxed one assertion while preserving reconstruction correctness.
* **Documentation**
  * Updated module documentation to describe the new fast‑kernel single‑pass backend.

[Review Change Stack](https://app.coderabbit.ai/change-stack/structured-world/structured-zstd/pull/216?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: zstd/src/encoding/match_generator.rs [hunk 1]
```diff
@@ -837,7 +848,16 @@ impl Matcher for MatchGeneratorDriver {
...
+                    // Donor level-1 Fast cParams: hash_log=14, mls=7.
+                    // window_log is derived from the resolved
+                    // LevelParams above (line ~771). Future per-level
+                    // table extension (hash_log scaled for small
+                    // sources) lands separately.
+                    MatcherStorage::Simple(FastKernelMatcher::with_params(
+                        params.window_log,
+                        FAST_LEVEL_1_HASH_LOG,
+                        FAST_LEVEL_1_MLS,
+                    ))
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/match_generator.rs:944]: @coderabbitai Acknowledged as KNOWN LIMITATION for phase 1b — added explicit doc comment at the FastKernelMatcher construction site (commit 4d4c581c) pointing to phase 3 (#198 items 2/3/5: 4-cursor ip0/ip1/ip2/ip3 lookahead + per-level kSearchStrength dispatch) where the acceleration gradient gets restored. Phase 1b is the donor-shape foundation; the gradient builds on top.

File: zstd/src/encoding/match_generator.rs [hunk 2]
```diff
@@ -1074,24 +1091,32 @@ impl Matcher for MatchGeneratorDriver {
...
+                let post = m.history_len_for_eviction_accounting();
+                // `accept_data` stashes the buffer in `pending`
+                // WITHOUT extending history — the eviction trigger
+                // fires on the actual `start_matching` /
+                // `skip_matching_with_hint` call. For accounting we
+                // therefore see `pre == post` here today; the
+                // budget retire / trim loop downstream queries
+                // `trim_to_window` directly when it needs the
+                // drained-byte delta.
+                evicted_bytes += pre.saturating_sub(post);
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/match_generator.rs:1217]: @coderabbitai Fixed in 2b231328 — moved the eviction into accept_data so the driver's commit_space pre/post history.len() delta correctly reports drained bytes to retire_dictionary_budget. Driver-site doc comment added in 4d4c581c.

File: zstd/src/encoding/simple/fast_matcher.rs [hunk 3]
```diff
@@ -0,0 +1,1095 @@
...
+            if drop_n > 0 {
+                self.history.drain(..drop_n);
+                self.prefix_start_index = self.prefix_start_index.saturating_add(drop_n as u32);
+                // The hash table holds ABSOLUTE positions into the
+                // pre-drain history; after draining, those positions
+                // point at evicted bytes. Clearing forces a fresh
+                // population from the new block onward — donor pays
+                // the same cost on `ZSTD_window_clear` after a stale
+                // reset.
+                self.hash_table.clear();
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/simple/fast_matcher.rs]: @coderabbitai Fixed in 661d6c79 — drain now (1) rebases prefix_start_index back to 1 instead of cumulative saturating_add and (2) rehashes the retained tail via prime_hash_table_for_range(0). Without rebase, accumulated saturating_add eventually rejected every match candidate; without rehash, the retained window would be 'dead history' as you noted. Both drain sites updated.

File: zstd/src/encoding/simple/fast_matcher.rs [hunk 4]
```diff
@@ -0,0 +1,1095 @@
...
+        if self.history.len() <= self.max_window_size {
+            return 0;
+        }
+        let drop_n = self.history.len() - self.max_window_size;
+        self.history.drain(..drop_n);
+        self.prefix_start_index = self.prefix_start_index.saturating_add(drop_n as u32);
+        // Hash table holds absolute positions into the pre-drain
+        // history — clear them as in the in-loop eviction path.
+        self.hash_table.clear();
+        drop_n
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/simple/fast_matcher.rs]: @coderabbitai Fixed in 5130d1e9 — last_block_start now saturating_sub by drop_n inside trim_to_window so post-trim last_committed_space() slices into the new history coordinate space.

File: zstd/src/encoding/match_generator.rs [hunk 5]
```diff
@@ -565,18 +588,30 @@ impl MatchGeneratorDriver {
...
+    /// `get_next_space()` call can reuse the allocation instead of
+    /// allocating fresh. Without this, the Simple backend would
+    /// allocate a new `Vec<u8>` per block — a measurable hot-path
+    /// cost when blocks are small (~128 KiB) and processed at
+    /// hundreds of MiB/s.
+    fn recycle_simple_space(&mut self) {
+        let slice_size = self.slice_size;
+        if let Some(mut space) = self.simple_mut().take_recycled_space() {
+            space.resize(space.capacity().max(slice_size), 0);
+            self.vec_pool.push(space);
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/match_generator.rs:662]: @coderabbitai Fixed in cea30a65 — went even further than your diff suggestion and just leave `len = 0` instead of `resize(slice_size, 0)`. `get_next_space()` already handles len normalization on pop (truncates if too long, resizes if too short), so the pool can carry buffers at any length without correctness impact. Saves the zero-fill entirely instead of just bounding it.

File: zstd/src/encoding/simple/fast_matcher.rs [hunk 6]
```diff
@@ -0,0 +1,1512 @@
...
+        let new_real_total = real_len.saturating_add(space.len());
+        let cap = self.max_window_size.saturating_mul(2);
+        if new_real_total > cap {
+            // Drop the OLDEST `drop_n` real bytes — those live at
+            // indices [RESERVED_PREFIX_BYTES .. RESERVED_PREFIX_BYTES + drop_n).
+            // The dummy at indices [0..RESERVED_PREFIX_BYTES) is
+            // preserved across the drain so the sentinel-0 invariant
+            // survives.
+            let drop_n = real_len.saturating_sub(self.max_window_size);
+            if drop_n > 0 {
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/simple/fast_matcher.rs:370]: @coderabbitai Fixed in d6290de2 — retain_real now derived from incoming block size: `cap.saturating_sub(space.len()).min(max_window_size)`. When the block alone exceeds cap, retain_real=0 (no historical context kept). Regression test in commit 8f14ecff verifies the 2× bound holds for oversize blocks. The debug_assert on `space.len() <= cap` from your suggestion isn't added because the math now degrades gracefully (truncate-to-zero context) rather than producing wrong output — but the doc comment in accept_data calls out that the cap holds 'as tightly as possible without truncating the caller's block', so callers committing > cap know what they're getting.

File: zstd/src/encoding/simple/fast_matcher.rs [hunk 7]
```diff
@@ -0,0 +1,1512 @@
...
+                    .last_block_start
+                    .saturating_sub(drop_n)
+                    .max(RESERVED_PREFIX_BYTES);
+                // Rehash retained tail (starting at the first real
+                // byte) so block N+1 can find matches against the
+                // bytes we explicitly kept. Without this, the
+                // retained window becomes "dead history" — visible
+                // in `history` but un-lookupable. Amortised over
+                // max_window_size of input it's O(1) per byte.
+                self.prime_hash_table_for_range(RESERVED_PREFIX_BYTES);
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/simple/fast_matcher.rs]: @coderabbitai Fixed in d6290de2 — extracted `drain_real_prefix(drop_n)` helper covering the full RESERVED-aware bookkeeping (drain + prefix_start_index rebase + table clear + last_block_start clamp + rehash). accept_data's eager-eviction branch and trim_to_window both call it. Helper docstring lists all five invariant preservations explicitly so the next drain-related change can't accidentally split them again.

**Comment by polaz** [comment]
Closing for clean review cycle — squashed 24 commits across review rounds 1-9 into a single foundation commit and force-pushed. New PR will open against the squashed history so review tooling (Copilot/CodeRabbit) gets a fresh pass without the noise of round-by-round fix commits.

All 21 review threads from this iteration were resolved before close. Substance is preserved in the squashed commit's body.
